sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/EnterCodeViewModel.kt

```diff
@@ -83,34 +92,60 @@ class EnterCodeViewModel @Inject constructor(
-                    }
-                    CONNECT_FAILED.code -> {
-                        showError(R.string.sync_connect_generic_error, result.reason)
-                    }
-                    CREATE_ACCOUNT_FAILED.code -> {
-                        showError(R.string.sync_create_account_generic_error, result.reason)
-                    }
-                    INVALID_CODE.code -> {
-                        viewState.value = viewState.value.copy(authState = AuthState.Error)
-                    }
-                    else -> {}
-                }
+                processError(result, pastedCode)
             }
         }
     }
 
-    private suspend fun showError(
-        message: Int,
-        reason: String,
-    ) {
-        command.send(ShowError(message = message, reason = reason))
+    private suspend fun processError(result: Error, pastedCode: String) {
+        if (result.code == ALREADY_SIGNED_IN.code && syncFeature.seamlessAccountSwitching().isEnabled()) {
+            command.send(AskToSwitchAccount(pastedCode))
+        } else {
+            if (result.code == INVALID_CODE.code) {
+                viewState.value = viewState.value.copy(authState = AuthState.Error)
+                return
+            }
+
+            when (result.code) {
+                ALREADY_SIGNED_IN.code -> R.string.sync_login_authenticated_device_error
+                LOGIN_FAILED.code -> R.string.sync_connect_login_error
+                CONNECT_FAILED.code -> R.string.sync_connect_generic_error
+                CREATE_ACCOUNT_FAILED.code -> R.string.sync_create_account_generic_error
+                INVALID_CODE.code -> R.string.sync_invalid_code_error
+                else -> null
+            }?.let { message ->
+                command.send(
+                    ShowError(
+                        message = message,
+                        reason = result.reason,
+                    ),
+                )
+            }
+        }
+    }
+
+    fun onUserAcceptedJoiningNewAccount(encodedStringCode: String) {
+        viewModelScope.launch(dispatchers.io()) {
+            val result = syncAccountRepository.logoutAndJoinNewAccount(encodedStringCode)
+            if (result is Error) {
+                when (result.code) {
+                    ALREADY_SIGNED_IN.code -> R.string.sync_login_authenticated_device_error
+                    LOGIN_FAILED.code -> R.string.sync_connect_login_error
+                    CONNECT_FAILED.code -> R.string.sync_connect_generic_error
+                    CREATE_ACCOUNT_FAILED.code -> R.string.sync_create_account_generic_error
+                    INVALID_CODE.code -> R.string.sync_invalid_code_error
+                    else -> null
+                }?.let { message ->
+                    command.send(
+                        ShowError(message = message, reason = result.reason),
+                    )
+                }
+            } else {
+                syncPixels.fireLoginPixel()
```

Comment:
this is the same pixel used elsewhere... just checking that's desirable having a new one specific to this use case?

Reply:
Yes, the actual user action is a login on a sync account, and we should signal it as that. To have some metrics around switching accounts, in a more specific ways, the metrics are added in https://github.com/duckduckgo/Android/pull/5283. So we will monitor both events 👍